strawberryfields/apps/train/parametrization.py

```diff
@@ -0,0 +1,406 @@
+
+    Args:
+        A (array): the adjacency matrix
+        sample (array): the sample
+
+    Returns:
+        float: the probability
+    """
+    n = len(A)
+    O = _Omat(A)
+    sample_big = np.hstack([sample, sample]).astype("int")
+    O_sub = reduction(O, sample_big)
+    scale = np.sqrt(np.linalg.det(np.identity(2 * n) - O))
+    return scale * tor(O_sub)
+
+
+def prob_photon_sample(A: np.ndarray, sample: np.ndarray) -> float:
+    """Calculate the probability of a sample of photon counts.
+
+    The input adjacency matrix must have singular values not exceeding one. This can be achieved
+    by rescaling an arbitrary adjacency matrix using :func:`rescale_adjacency`.
+
+    Args:
+        A (array): the adjacency matrix
+        sample (array): the sample
+
+    Returns:
+        float: the probability
+    """
+    n = len(A)
+    mu = np.zeros(2 * n)
+    cov = A_to_cov(A)
+    sample = np.array(sample, dtype="int")
+    return np.abs(pure_state_amplitude(mu, cov, sample, hbar=sf.hbar)) ** 2
+
+
+def A_to_cov(A: np.ndarray) -> np.ndarray:
+    """Convert an adjacency matrix to a covariance matrix of a GBS device.
+
+    The input adjacency matrix must have singular values not exceeding one. This can be achieved
+    by rescaling an arbitrary adjacency matrix using :func:`rescale_adjacency`.
+
+    **Example usage:**
+
+    >>> a = np.ones((3, 3))
+    >>> a = rescale_adjacency(a, 2, True)
+    >>> cov = A_to_cov(a)
+
+    Args:
+        A (array): the adjacency matrix
+
+    Returns:
+        array: the covariance matrix of :math:`A`
+    """
+    n = len(A)
+    I = np.identity(2 * n)
+    return sf.hbar * (np.linalg.inv(I - _Omat(A)) - I / 2)
+
+
+class VGBS:
+    r"""Create a variational GBS model for optimization and machine learning.
+
+    An input adjacency matrix :math:`A` can be varied using:
+
+    .. math::
+
+        A(\theta) = W(\theta) A W(\theta),
+
+    with :math:`W` a diagonal matrix of weights that depend on a set of parameters :math:`\theta`.
+
+    By varying :math:`\theta`, the distribution of samples from GBS can be trained to solve
+    stochastic optimization and unsupervised machine learning problems.
+
+    The above variational model can be used in both the threshold and PNR modes of GBS, which
+    is specified using the ``threshold`` flag. An initial value for the mean number of clicks
+    (if ``threshold=True``) or mean number of photons (if ``threshold=False``) is also required.
+    The initial value ``n_mean`` should be set high since varying :math:`\theta` can only lower
+    the mean number of clicks or photons.
+
+    The mapping from :math:`\theta` to :math:`W(\theta)` is specified using the ``embedding``
+    argument. TODO mention embedding module
+
+    **Example usage:**
+
+    >>> g = nx.erdos_renyi_graph(4, 0.7, seed=1967)
+    >>> A = nx.to_numpy_array(g)
+    >>> embedding = Exp(4)  # TODO does this need updating?
```

Comment:
as per the comment, will need an update alike to `embedding = ExpFeatures([[1,2,3,4]])` (`[[1,2,3,4]]` is just example features)

Reply:
I updated to `>>> embedding = train.embed.Exp(4)` as it's simpler to go with the `Exp` embedding in general. Do you think `train.embed` is the best way to point to it though?